packages/flutter/lib/src/material/button.dart

```diff
@@ -231,61 +240,82 @@ class RawMaterialButton extends StatefulWidget {
 }
 
 class _RawMaterialButtonState extends State<RawMaterialButton> {
-  bool _highlight = false;
-  bool _focused = false;
-  bool _hovering = false;
+  final Set<MaterialState> _states = <MaterialState>{};
+
+  bool get _hovered => _states.contains(MaterialState.hovered);
+  bool get _focused => _states.contains(MaterialState.focused);
+  bool get _pressed => _states.contains(MaterialState.pressed);
+  bool get _disabled => _states.contains(MaterialState.disabled);
+
+  void _updateState(MaterialState state, bool value) {
+    value ? _states.add(state) : _states.remove(state);
+  }
 
   void _handleHighlightChanged(bool value) {
-    if (_highlight != value) {
+    if (_pressed != value) {
       setState(() {
-        _highlight = value;
+        _updateState(MaterialState.pressed, value);
         if (widget.onHighlightChanged != null) {
           widget.onHighlightChanged(value);
         }
       });
     }
   }
 
+  @override
+  void initState() {
+    super.initState();
+    _updateState(MaterialState.disabled, !widget.enabled);
+  }
+
   @override
   void didUpdateWidget(RawMaterialButton oldWidget) {
     super.didUpdateWidget(oldWidget);
-    if (_highlight && !widget.enabled) {
-      _highlight = false;
```

Comment:
Is this not necessary anymore?

Reply:
Oh, is it resolved by clearing the states below?

Reply:
Removed.